src/gsy_e/models/strategy/energy_parameters/heatpump/heat_pump.py

```diff
@@ -76,6 +121,108 @@ def tanks(self) -> AllTanksEnergyParameters:
+            heat_demand_kW=min_heat_demand_kW,
+        )
+
+        if cop == 0:
+            return 0
+        min_energy_consumption_kWh = min_heat_demand_kW / cop
+
+        if min_energy_consumption_kWh > self._max_energy_consumption_kWh:
+            return self._max_energy_consumption_kWh
+        return min_energy_consumption_kWh
+
+    def increase_tank_temp_from_bought_energy(
+        self, time_slot: DateTime, bought_energy_kWh: float, source_temp_C: float
+    ):
+        """
+        Increase the tanks temperature from the energy that was bought by the heat pump strategy.
+        """
+        traded_heat_energy_kJ = self._calc_Q_kJ_from_energy_kWh(
+            time_slot, bought_energy_kWh, source_temp_C
+        )
+        self._charger.charge(traded_heat_energy_kJ, time_slot)
+
+    def decrease_tank_temp_from_heat_demand(self, heat_demand_kJ: float, time_slot: DateTime):
+        """Decrease the tanks temperature according to the heat demand."""
+        self._charger.discharge(heat_demand_kJ, time_slot)
+
+    def update_cop(self, heat_demand_J: float, source_temp_C: float, time_slot: DateTime):
+        """Update the COP of the heat pump in its state class."""
+        cop = self._calc_cop(heat_demand_J, source_temp_C, time_slot)
+        self._hp_state.set_cop(time_slot, cop)
+
+    def _calc_cop(
+        self, heat_demand_Q_J: float, source_temp_C: float, time_slot: DateTime
+    ) -> float:
+        """
+        Return the coefficient of performance (COP) for a given ambient and storage temperature.
+        The COP of a heat pump depends on various parameters, but can be modeled using
+        the two temperatures.
+        Generally, the higher the temperature difference between the source and the sink,
+        the lower the efficiency of the heat pump (the lower COP).
+        """
+        # 1 J = 1 W s
+        heat_demand_kW = (
+            heat_demand_Q_J / self._slot_length.total_seconds() / 1000 if heat_demand_Q_J else None
+        )
+        return self._cop_model.calc_cop(
+            source_temp_C=source_temp_C,
+            condenser_temp_C=self._charger.get_condenser_temperature_C(time_slot),
+            heat_demand_kW=heat_demand_kW,
+        )
+
+    def _calc_Q_kJ_from_energy_kWh(
+        self, time_slot: DateTime, energy_kWh: float, source_temp_C: float
+    ) -> float:
+        condenser_temp = self._charger.get_condenser_temperature_C(time_slot)
+        cop = self._cop_model.calc_cop(source_temp_C, condenser_temp, energy_kWh)
+        return convert_kWh_to_kJ(cop * energy_kWh)
+
+    def _calc_energy_kWh_from_Q_kJ(self, time_slot: DateTime, Q_energy_kJ: float) -> float:
+        cop = self._hp_state.get_cop(time_slot)
+        if cop == 0:
+            return 0
+        return convert_kJ_to_kWh(Q_energy_kJ / self._hp_state.get_cop(time_slot))
+
+    def _last_time_slot(self, current_market_slot: DateTime) -> DateTime:
```

Comment:
I do not see this being used anywhere.

Reply:
This is a remnant from an old implementation, removing.